### ISSUE TITLE
Optional CPU limit in while: True loops

### ISSUE THREAD

**Issue by otter-in-a-suit** [abhiTronix/vidgear#107]
As vidgear uses several, non-limited `while True` loops, I can see 100% CPU usage per core, most notably in `Netgear.recv()`. Includes fans going at 100%.

## Detailed Description

While I'm sure there are many dependencies and special cases I'm ignoring, my fork currently uses a simple sleep(0.02) as such: [https://github.com/otter-in-a-suit/vidgear/commit/a8534de14f0977d5e507d7a7afdf1f0d877751ab](fork)

I'm not sure if there is a better, synchronized way, but any form of limiter would be amazing.

### Context

I am using Vidgear for a network-driven security camera: https://github.com/otter-in-a-suit/scarecrow

Running vanilla vidgear causes 100% CPU usage on all cores used by the API.

### Your Environment

* VidGear version: 0.1.6 (fork)
* Branch: Development
* Python version: 3.7.3
* pip version: 18.1
* Operating System and version: Linux pop-os 5.3.0-7625-generic #27~1576774560~19.04~f432cd8-Ubuntu SMP Fri Jan 10 16:25:52 UTC  x86_64 x86_64 x86_64 GNU/Linux

### Any Other Important Information

With 0.02s sleep:
![with_limit](https://user-images.githubusercontent.com/6397962/74591329-019cb480-500f-11ea-8a10-96a22c65440c.png)

Without:
![without_limit](https://user-images.githubusercontent.com/6397962/74591331-02354b00-500f-11ea-995f-1fd1132ca2cd.png)

Both on an empty receive, i.e. with no active sender.

**Comment by abhiTronix** [comment]
@otter-in-a-suit Why not add `time.sleep(0.02)` in your script code rather in vidgear code? Isn't it will be more convenient? It will have the same effect.

**Comment by otter-in-a-suit** [comment]
Please correct me if i read the code wrong, but from what I can tell, self.terminate in the Neatgear.recv() method evaluates to a while True, as without incoming data (which is a use case, I believe), it waits forever on data and hence, never sets the self.terminate flag.

**Comment by abhiTronix** [comment]
> self.terminate in the Neatgear.recv() method evaluates to a while True, as without incoming data (which is a use case, I believe), it waits forever on data and hence, never sets the self.terminate flag.

Not exactly, `self.terminate` flag depends on two factors in receive mode(i.e `receive_mode = True`):

1. **`terminate_flag` received from the server to which it is connected:**

 Makes `self.terminate =True`. See here:

https://github.com/abhiTronix/vidgear/blob/8357aa5d6df59df45bd21a7c23c0950c6b152f59/vidgear/gears/netgear.py#L554-L576

2. **User clicks `CTRL+C` keyboard Interrupt on the keyboard:**

This will force terminates the loop by making `self.terminate =True`. See here:

https://github.com/abhiTronix/vidgear/blob/8357aa5d6df59df45bd21a7c23c0950c6b152f59/vidgear/gears/netgear.py#L745-L760

**Comment by abhiTronix** [comment]
@otter-in-a-suit The termination based on Server response has both pros and cons. But you can yourself add a timer in your script to tackle this so that it self-terminates after a fixed amount of time if no response is received from the server.

**Comment by otter-in-a-suit** [comment]
Thanks for the explanation! 

In the use case I am referring to, the server might no exists (yet). This could happen in a couple of scenarios, I believe. For instance, if the device sending the stream is not available, busy, has bad network connection, or simply does not send data continuously. An example for the latter could be a doorbell-camera, which would only send data upon somebody pressing it. In these cases, I would still want my server to be listening, but not at 100% CPU capacity.

If I add a thread that waits for X seconds on a response, I am still wasting a lot of CPU cycles for the thread, the server will run on 100% CPU on one core while waiting during that time, I add artificial delay in the times I'm not actively listening, and I am forced to re-start the receiver after a while, as I do expect a response eventually. In this scenario, there also won't be any user input, as it's running headless, for instance via a systemd service.

**Comment by abhiTronix** [comment]
> but not at 100% CPU capacity.

@otter-in-a-suit Something is wrong here. Can you share your script code that you're using to access NetGear API?

**Comment by otter-in-a-suit** [comment]
Here's a minimal example that I stripped from here: https://github.com/otter-in-a-suit/scarecrow/blob/master/server/receiver.py#L56

```
from vidgear.gears import NetGear
import logging
import cv2

logger = logging.getLogger()
client = NetGear(address='127.0.0.1', port=str(5454), protocol='tcp',
                     pattern=1, receive_mode=True, logging=True)


while True:
    # receive frames from network
    frame = client.recv()
    logger.debug('Image received')
    #image_np = np.copy(frame)
    # check if frame is None
    if frame is None:
        logger.error('No frame available')
        break

    key = cv2.waitKey(1) & 0xFF
    # check for 'q' key-press
    if key == ord("q"):
        # if 'q' key-pressed break out
        break
```

Please notice how there is no sender active, for the use case I've outline above.

Without 0.02 sleep:
![Screenshot from 2020-02-17 18-48-03](https://user-images.githubusercontent.com/6397962/74692683-3c097b80-51e0-11ea-91d4-dccd0cfe5b46.png)

With:
![Screenshot from 2020-02-17 18-47-21](https://user-images.githubusercontent.com/6397962/74692688-3e6bd580-51e0-11ea-9129-0f33f93f55a0.png)

See how one core is at 100%. The fans on my System76 Gazelle are also going crazy due to that.

Hope this helps!

**Comment by abhiTronix** [comment]
> address='127.0.0.1'

@otter-in-a-suit This means you're running your both server & client on your local machine, and both are sharing the same resources, thereby 100% CPU is obvious and is certainly not the correct way to benchmark it. But yeah, NetGear is surely using a lot of resources(not 100%) but more then it is intended to, and also ZMQ internals are also responsible for the same.

No doubt, Adding `time.sleep()` is suspending CPU for a short duration, but it is just a dirty fix but not an ideal solution for doing the same, you're also delaying your output on a faster connection. **The correct solution that I'm planning to incorporate [`asyncio`](https://docs.python.org/3/library/asyncio.html) with vidgear for a long time and I think it is the future for this library. Ascynio has many inbuilt functions that can quadruple performance while using minimum system resources in NetGear. I'm postponing this feature for the future since I'm super busy with other projects but when I'll push PR related to it you'll be informed first.** So stay tuned.

**Comment by otter-in-a-suit** [comment]
The same thing happens over the network. But as stated in the original post, yes, sleep is a really bad workaround.

Thanks for your explanation - looking forward to any improvements! :)

**Comment by abhiTronix** [comment]
**Hey @otter-in-a-suit, Thanks for your patience, Kindly reinstall the new [NetGear_Async API](https://github.com/abhiTronix/vidgear/wiki/NetGear_Async) PR(#114) as follows:**

```sh
git clone https://github.com/abhiTronix/vidgear.git
cd vidgear
git checkout development
sudo pip3 install .
cd
```

and **test the usage examples from [its wiki](https://github.com/abhiTronix/vidgear/wiki/NetGear_Async#usage) and see if it solves memory problem?** and then revert here!

**Comment by abhiTronix** [comment]
**@otter-in-a-suit Closed and marked solved for now. Kindly share your results when you're ready. Good luck!**

**Comment by otter-in-a-suit** [comment]
# Overall
Thank you for this change! it will take me a while to update my code to use `asyncio`, but I have run several preliminary tests with the new branch.

I haven't pushed my changes to my own project just yet.

# Test approach:

- Updated https://github.com/otter-in-a-suit/vidgear/tree/development from master repo
- Updated https://github.com/otter-in-a-suit/scarecrow/blob/development/sbin/install_vidgear.sh
- Ran ./sbin/install_vidgear.sh
- Check
```
(env) christian @ pop-os ➜  scarecrow git:(testing) ✗  pip3 list | grep vidgear
vidgear              0.1.7.dev1 
```
- Added `asyncio` to `requirements.txt` / `setup.py`
- Updated `sender`, `receiver` to new async code


# Problems & Resolutions

## Dependencies
### Issue:
- `E   ModuleNotFoundError: No module named 'starlette'`
- `E   ModuleNotFoundError: No module named 'aiofiles'`
- `E   ModuleNotFoundError: No module named 'aiohttp'`
...

### Background:
You are using `pip3 install .[asyncio]`, which references your setup.py. 

### Resolution:
As this contains these dependencies, I've added the line to my `sbin` script. Maybe update the docs?


### Issue:
```
NetGear_Async :: ERROR :: 'NetGear_Async' object has no attribute '_NetGear_Async__address'
Traceback (most recent call last):
  File "/home/christian/workspace/scarecrow/env/lib/python3.7/site-packages/vidgear/gears/asyncio/netgear_async.py", line 304, in __server_handler
    self.__protocol + "://" + str(self.__address) + ":" + str(self.__port)
```

### Background:
If `receive_mode` is set to False, `__address` is not set. This is an oversight in the else clause.

### Resolution:
**Pull request:** https://github.com/abhiTronix/vidgear/pull/118

### Issue:
```NetGear_Async :: WARNING :: Invalid pattern. Defaulting to `zmq.PAIR`!```

### Background:
Your example in the docs does not show all kwargs. When updating from legacy to `0.1.7.dev1`, old code breaks. This should return the invalid pattern. 


### Resolution:
**Pull request:** https://github.com/abhiTronix/vidgear/pull/118

### Issue:
On any of the above failures, the thread does not complete and deadlocks.

### Background:
I am not familiar with `asyncio` (yet :)). Probably an issue on my end.

### Resolution:
N/A

**Comment by abhiTronix** [comment]
> Dependencies
> Issue:
> 
>     E ModuleNotFoundError: No module named 'starlette'
>     E ModuleNotFoundError: No module named 'aiofiles'
>     E ModuleNotFoundError: No module named 'aiohttp'
>     ...
> 
> Background:
> 
> You are using pip3 install .[asyncio], which references your setup.py.
> Resolution:
>As this contains these dependencies, I've added the line to my sbin script. Maybe update the docs?

@otter-in-a-suit Sorry I didn't added line `pip3 install .[asyncio]` above. **But you can already find correct way to install dependencies in [its wiki](https://github.com/abhiTronix/vidgear/wiki/NetGear_Async#requirements)**

> 
> Issue:
> 
> NetGear_Async :: WARNING :: Invalid pattern. Defaulting to `zmq.PAIR`!
> Background:
> 
> Your example in the docs does not show all kwargs. When updating from legacy to 0.1.7.dev1, old code breaks. This should return the invalid pattern.
> Resolution:
> 
> Pull request: #118 
Let's work on this issue on other issue you created more recently.

**Comment by abhiTronix** [comment]
@otter-in-a-suit adding small sleep command does helps in reducing load. I'll add this fix in next PR.

**Comment by abhiTronix** [comment]
**Successfully resolved and merged in commit 8f8d951**
